feat(scan): define base abstractions to manage scans

ostridm opened the pull request:
closes #42

ostridm committed:
feat(scan): define base abstractions to manage scans

closes #42

derevnjuk commented on src/SecTester.Scan/Har/HarContent.cs:
```diff
@@ -0,0 +1,5 @@
+namespace SecTester.Scan.Har;
+
+public class HarContent
```
Use `record` instead
```suggestion
public record Har

```

derevnjuk commented on src/SecTester.Scan/Har/UploadHarOptions.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace SecTester.Scan.Har;
+
+public class UploadHarOptions
```
ditto

derevnjuk commented on src/SecTester.Scan/Har/UploadHarOptions.cs:
```diff
@@ -0,0 +1,20 @@
+  public string FileName { get; set; }
+  public bool? Discard { get; set; }
+
+  public UploadHarOptions(HarContent harContent, string fileName, bool? discard = default)
```
Use primary constructor instead

```csharp
public record UploadHarOptions(HarContent HarContent, string FileName, bool? Discard = default) {
  public object HarContent { get; set; } = HarContent ?? throw new ArgumentNullException(nameof(HarContent));

  [JsonPropertyName("filename")] 
  public string FileName { get; set; } = FileName ?? throw new ArgumentNullException(nameof(FileName));

  public bool? Discard { get; set; } = Discard;
}
```

derevnjuk commented on src/SecTester.Scan/Models/AttackParamLocation.cs:
```diff
@@ -0,0 +1,27 @@
+  [EnumMember(Value = "artifical-query")]
+  ArtificalQuery,
+  
+  [EnumMember(Value = "body")] 
```
Use `EnumMember` only when it is required, instead add converters `new JsonStringEnumConverter(JsonNamingPolicy.CamelCase, false)` in our `MessageSerializer` or globally. Yet, you have to implement a `snake_case` naming policy.

derevnjuk commented on src/SecTester.Scan/Models/Comment.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace SecTester.Scan.Models;
+
+public class Comment
```
ditto

derevnjuk commented on src/SecTester.Scan/Har/UploadHarOptions.cs:
```diff
@@ -0,0 +1,20 @@
+
+public class UploadHarOptions
+{
+  public HarContent HarContent { get; set; }
```
Use init only setter in immutable types

```suggestion
  public HarContent HarContent { get; init; }

```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+    new(new[] { ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued });
+
+  private static readonly HashSet<ScanStatus> DoneStatuses =
+    new(new[] { ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped });
```
```suggestion
    new() { ScanStatus.Disrupted, ScanStatus.Done, ScanStatus.Failed, ScanStatus.Stopped };

```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+public class Scan : IDisposable
+{
+  private static readonly HashSet<ScanStatus> ActiveStatuses =
+    new(new[] { ScanStatus.Pending, ScanStatus.Running, ScanStatus.Queued });
```
ditto

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+
+public class Scan : IDisposable
+{
+  private static readonly HashSet<ScanStatus> ActiveStatuses =
```
The collection must be `IReadOnlyCollection`
```suggestion
  private static readonly IReadOnlyCollection<ScanStatus> ActiveStatuses =

```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+    GC.SuppressFinalize(this);
+  }
+
+  public Task<Issue[]> Issues(CancellationToken token = default)
```
Expose `IEnumerable` instead of `Array` 

```suggestion
  public Task<IEnumerable<Issue>> Issues(CancellationToken token = default)

```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+    throw new NotImplementedException();
+  }
+
+  public IAsyncEnumerable<ScanState> Status()
```
You should not accept `CancellationToken` when there is a side-effect that may lead to inconsistency, use it only with callbacks or in async loops. This is exactly the case.

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -0,0 +1,53 @@
+    GC.SuppressFinalize(this);
+  }
+
+  public Task<Issue[]> Issues(int? limit, string? nextId, DateTime? nextCreatedAt)
```
Currently, we don't support server-side pagination
```suggestion
  public Task<Issue[]> Issues()

```

derevnjuk replied:
Expose `IEnumerable` instead of `Array` in public API

```suggestion
  public Task<IEnu> Issues(int? limit, string? nextId, DateTime? nextCreatedAt)

```

derevnjuk commented on src/SecTester.Scan/ScanFactory.cs:
```diff
@@ -0,0 +1,9 @@
+
+public interface ScanFactory
+{
+  Task<Scan> CreateScan(ScanSettingsOptions settingsOptions, ScanOptions options, CancellationToken token = default);
```
See comment above

```suggestion
  Task<Scan> CreateScan(ScanSettingsOptions settingsOptions, ScanOptions options);

```

derevnjuk commented on src/SecTester.Scan/Scans.cs:
```diff
@@ -0,0 +1,18 @@
+public interface Scans
+{
+  Task<string> CreateScan(ScanConfig config);
+  Task<List<Issue>> ListIssues(string id, int? limit, string? nextId, DateTime? nextCreatedAt);
```
```suggestion
  Task<IEnumerable<Issue>> ListIssues(string id, int? limit, string? nextId, DateTime? nextCreatedAt);

```

derevnjuk replied:
```suggestion
  Task<List<Issue>> ListIssues(string id);

```

derevnjuk commented on src/SecTester.Scan/Target/TargetOptions.cs:
```diff
@@ -0,0 +1,34 @@
+  /// <summary>
+  /// The request method to be used when making the request, GET by default
+  /// </summary>
+  RequestMethod? Method { get; set; }
```
```suggestion
  HttpMethod? Method { get; set; }

```

derevnjuk commented on src/SecTester.Scan/Target/TargetOptions.cs:
```diff
@@ -0,0 +1,34 @@
+  /// <summary>
+  /// The headers
+  /// </summary>
+  Dictionary<string, string>? Headers { get; set; }
```
```suggestion
  IDictionary<string, IEnumerable<string>>? Headers { get; set; }

```

derevnjuk commented on src/SecTester.Scan/Target/TargetOptions.cs:
```diff
@@ -0,0 +1,34 @@
+  /// <summary>
+  /// The query parameters to be sent with the request
+  /// </summary>
+  Dictionary<string, string>? Query { get; set; }
```
ditto

derevnjuk commented on src/SecTester.Scan/Target/TargetOptions.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace SecTester.Scan.Target;
+
+public interface TargetOptions
```
It is pointless to declare an interface, I hardly imagine that the user will implement it. Also, exposing a setter make it impossible to implement an immutable class

derevnjuk replied:
You should provide the ability to pass a custom query serializer (e.g. `Func<IDictionary<string, IEnumerable<string>>, string> SerializeQuery`)

derevnjuk reviewed (changes requested):
As we discussed yesterday, there is not a reason to write noop tests for POJO classes without logic. Please remove them.

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023982848
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023988118
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1024005614
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023981663
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023981012
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023979909
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023979085
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023969897
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023966613
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023965095
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023965004
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023921274
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023921837
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023924067
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023925579
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023926331
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023926331
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1024141527

closes #42

derevnjuk committed:
refactor(scan): clean up code

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023925366

closes #42

derevnjuk commented on src/SecTester.Core/Utils/StringUtils.cs:
```diff
@@ -0,0 +1,66 @@
+
+namespace SecTester.Core.Utils;
+
+public static class StringUtils
```
Why is it so complex?) It can be written in a few lines of code

```csharp
public static class StringUtils
{
  private static readonly Regex CamelCaseRegex = new(@"([a-z0-9])([A-Z])");
  private static readonly Regex PascalCaseRegex = new(@"([A-Z])([A-Z][a-z])");

  private static readonly Regex[] SupportedCaseRegexes =
  {
    PascalCaseRegex, CamelCaseRegex
  };

  public static string? ToSnakeCase(this string? value)
  {
    return string.IsNullOrEmpty(value)
      ? value
      : SupportedCaseRegexes.Aggregate(value!,
          (input, regex) => regex.Replace(input, "$1_$2"))
        .ToLowerInvariant();

  }
}
```

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -0,0 +1,33 @@
+#pragma warning disable CS8601
```
Please don't use pragma

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -0,0 +1,33 @@
+public class StringUtilsTests
+{
+  public static readonly object[][] SnakeCaseMapping = {
+    new object[] { null as object, null as object },
+    new object[] { "", "" },
+    new object[] { "0", "0" },
+    new[] { "FOO", "foo" }, 
+    new[] { "Bar", "bar" }, 
+    new[] { "fooBar", "foo_bar" },
+    new[] { "FooBar", "foo_bar" },
+    new[] { "1FooBar", "1_foo_bar" },
+    new[] { "Foo1Bar", "foo1_bar" },
+    new[] { "FooBar1", "foo_bar1" },
+    new[] { "Foo_Bar", "foo_bar" },
+    new[] { "Foo_bar", "foo_bar" },
+    new[] { "foo_Bar", "foo_bar" },
+    new[] { "foo_bar", "foo_bar" },
```
The test data can be as close as possible to really and DRY.

```suggestion
    new object[] { null!, null! },
    new object[] { "", "" },
    new object[] { "lowercase", "lowercase" }, 
    new object[] { "UPPERCASE", "uppercase" }, 
    new object[] { "camelCase", "camel_case" }, 
    new object[] { "snake_case", "snake_case" },
    new object[] { "kebab-case", "kebab-case" },
    new object[] { "PascalCase", "pascal_case" },
    new object[] { "camelCase_with_snake", "camel_case_with_snake" },
    new object[] { "PascalCase_with_snake", "pascal_case_with_snake" },
    new object[] { "kebab-case_with_snake", "kebab-case_with_snake" },

```

derevnjuk commented on test/SecTester.Bus.Tests/Dispatchers/DefaultMessageSerializerTests.cs:
```diff
@@ -65,4 +73,37 @@ public void Deserialize_GenericReturnType_ReturnsObject(string input)
+  }
+  
+  [Fact]
+  public void  Deserialize_GivenMissingFiledInput_ReturnObject()
```
```suggestion
  public void Deserialize_GivenMissingFiledInput_ReturnObject()
```

derevnjuk commented on test/SecTester.Bus.Tests/Dispatchers/DefaultMessageSerializerTests.cs:
```diff
@@ -65,4 +73,37 @@ public void Deserialize_GenericReturnType_ReturnsObject(string input)
+
+  [Theory]
+  [MemberData(nameof(EnumValues))]
+  public void  Deserialize_GivenString_ReturnEnumValue(Type type, object value, string valueString)
```
ditto

derevnjuk replied:
ditto

derevnjuk replied:
ditto
```suggestion
  public void Deserialize_GivenString_ReturnEnumValue(Type type, object value, string valueString)
```

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023926331
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1023986008
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1024160838
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1024160838
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026329023
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026329118
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026327038
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026325243
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026325996

closes #42

derevnjuk commented on src/SecTester.Scan/Models/Request.cs:
```diff
@@ -1,13 +1,14 @@
+  public IDictionary<string, string>? Headers { get; init; }
+  public string? Body { get; init; }
+  public Protocol? Protocol { get; init; }
+}
```
Why there are so many weird changes?

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -0,0 +1,67 @@
+
+public class JsonHttpMethodEnumerationStringConverter : JsonConverter<HttpMethod>
+{
+  private static readonly HttpMethod _patch = new("PATCH");
+  private static readonly HttpMethod _copy = new("COPY");
+  private static readonly HttpMethod _link = new("LINK");
+  private static readonly HttpMethod _unlink = new("UNLINK");
+  private static readonly HttpMethod _purge = new("PURGE");
+  private static readonly HttpMethod _lock = new("LOCK");
+  private static readonly HttpMethod _unlock = new("UNLOCK");
+  private static readonly HttpMethod _propfind = new("PROPFIND");
+  private static readonly HttpMethod _view = new("VIEW");
+  private static readonly HttpMethod _trace = new("TRACE");
```
You don't need to create static instances, `HttpMethod` implements an equality operator, and `IEquatable` as well. You can get rid of `_map` creating a new instance of `HttpMethod`

ostridm replied:
That map implicitly does verifiction of the incoming value for its  validity like an `Enum` type. Also it prevents deserializer from redundant object creation. I believe since the `HttpMethod` internally implemented relying on static memebers we may use the same approach.

ostridm replied:
Also, suggest to verify the `Method` coming from the client side and fail serialization to save roundtrip to server.

derevnjuk replied:
If so, why don't use reflection to avoid mistyping or any further changes:

```csharp
  private static readonly IEnumerable<HttpMethod> BaseMethods = typeof(HttpMethod)
    .GetProperties(BindingFlags.Public |
                   BindingFlags.Static |
                   BindingFlags.DeclaredOnly)
    .Select(f => f.GetValue(null))
    .Cast<HttpMethod>();

  private static readonly IEnumerable<HttpMethod> CustomMethods = new List<HttpMethod>
  {
    new("PATCH"),
    new("COPY"),
    new("LINK"),
    new("UNLINK"),
    new("PURGE"),
    new("LOCK"),
    new("UNLOCK"),
    new("PROPFIND"),
    new("VIEW")
  };
```

Then you can concat them and convert a result to a dictionary

```csharp
CustomMethods.Concat(BaseMethods).ToDictionary(x => x.ToString(), x => x);
```

However, you need to resolve the issue with duplicates (e.g. there is Patch in NET 6):

```csharp
private static IDictionary<string, HttpMethod> GetMethods()
{
  if (Methods is not { Count: 0 })
  {
    return Methods;
  }

  var methods = CustomMethods.Concat(BaseMethods);
    
  foreach (var method in methods)
  {
    Methods[method.Method] = method;
  }

  return Methods;
}
```
So, if they add new methods to HttpMethod, we will use them (i.e. `HttpMethod.Link.Should().BeSameAs(expected)`)

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+      case JsonTokenType.String:
+        var token = reader.GetString();
+        if (token is not null && _map.TryGetValue(token, out HttpMethod httpMethod))
```
```suggestion
        if (token is not null && _map.TryGetValue(token, out var httpMethod))
```

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        throw new JsonException(
+          $"Unexpected value {token} when parsing the {nameof(HttpMethod)}.");
```
Please inverse `if`

derevnjuk commented on src/SecTester.Bus/Internal/SnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Text.Json;
+using SecTester.Core.Utils;
+
+namespace SecTester.Core.Json;
```
Please make sure namespaces are aligned with folders
```suggestion
namespace SecTester.Bus.Internal;

```

derevnjuk commented on src/SecTester.Bus/Internal/SnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,14 @@
+
+class SnakeCaseNamingPolicy : JsonNamingPolicy
+{
+  public static SnakeCaseNamingPolicy Instance { get; } = new SnakeCaseNamingPolicy();
```
```suggestion
  public static JsonNamingPolicy Instance { get; } = new SnakeCaseNamingPolicy();

```

derevnjuk commented on test/SecTester.Bus.Tests/Dispatchers/DefaultMessageSerializerTests.cs:
```diff
@@ -65,4 +73,37 @@ public void Deserialize_GenericReturnType_ReturnsObject(string input)
+  
+  [Theory]
+  [MemberData(nameof(EnumValues))]
+  public void Serialize_GivenEnumValue_ReturnString(Type type, object value, string valueString)
```
The `type` parameter is not used (see warnings)

```suggestion
  public void Serialize_GivenEnumValue_ReturnString(object value, string valueString)
```

derevnjuk replied:
```suggestion
  public void Serialize_GivenEnumValue_ReturnString(Type type, object input, string expected)
```

derevnjuk commented on test/SecTester.Bus.Tests/Dispatchers/DefaultMessageSerializerTests.cs:
```diff
@@ -65,4 +73,37 @@ public void Deserialize_GenericReturnType_ReturnsObject(string input)
+  }
+  
+  [Fact]
+  public void  Deserialize_GivenMissingFieldInput_ReturnObject()
```
```suggestion
  public void Deserialize_GivenMissingFieldInput_ReturnObject()
```

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+
+  public static readonly IEnumerable<object[]> HttpMethodEnumerable = new List<object[]>()
+  {
+    new object[] { HttpMethod.Delete, @"""DELETE""", typeof(HttpMethod) },
```
The type is redundant for all cases, you can pass it within `Deserialize_GivenString_ReturnEnumValue`
```suggestion
    new object[] { HttpMethod.Delete, @"""DELETE""" },
```

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+{
+  private readonly DefaultMessageSerializer _sut = new();
+
+  public static readonly IEnumerable<object[]> HttpMethodEnumerable = new List<object[]>()
```
```suggestion
  public static readonly IEnumerable<object[]> HttpMethodEnumerable = new List<object[]>
```

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+
+  [Theory]
+  [MemberData(nameof(HttpMethodEnumerable))]
+  public void Serialize_GivenEnumValue_ReturnString(object value, string valueString, Type type)
```
ditto

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+
+  [Theory]
+  [MemberData(nameof(HttpMethodEnumerable))]
+  public void Deserialize_GivenString_ReturnEnumValue(object value, string valueString, Type type)
```
ditto

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+  }
+  
+  [Fact]
+  public void  Deserialize_GivenMissingFieldInput_ReturnObject()
```
```suggestion
  public void Deserialize_GivenMissingFieldInput_ReturnObject()
```

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -0,0 +1,29 @@
+
+public class StringUtilsTests
+{
+  public static readonly IEnumerable<object[]> SnakeCaseMapping = new List<object[]>(){
```
```suggestion
  public static readonly IEnumerable<object[]> SnakeCaseMapping = new List<object[]>
  {
```

derevnjuk commented on test/SecTester.Scan.Tests/Models/ModelEnumJsonSerializationTests.cs:
```diff
@@ -0,0 +1,152 @@
+{
+  private readonly DefaultMessageSerializer _sut = new();
+
+  public static readonly IEnumerable<object[]> AttackParamLocationEnumerable = new List<object[]>()
```
The same for all of them below

```suggestion
  public static readonly IEnumerable<object[]> AttackParamLocationEnumerable = new List<object[]>
```

derevnjuk commented on test/SecTester.Scan.Tests/Models/ModelEnumJsonSerializationTests.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace SecTester.Scan.Tests.Models;
+
+public class JsonStringEnumMemberConverterTests
```
The test suite name and SUT are different, move all of them to `DefaultMessageSerializerTests`

derevnjuk commented on test/SecTester.Bus.Tests/Internal/JsonHttpMethodEnumerationStringConverterTests.cs:
```diff
@@ -0,0 +1,59 @@
+namespace SecTester.Bus.Tests.Internal;
+
+public class JsonHttpMethodEnumerationStringConverterTests
```
The test suite name and SUT are different, move all of them to `DefaultMessageSerializerTests`

derevnjuk commented on src/SecTester.Bus/Internal/SnakeCaseNamingPolicy.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace SecTester.Core.Json;
+
+class SnakeCaseNamingPolicy : JsonNamingPolicy
```
```suggestion
internal class SnakeCaseNamingPolicy : JsonNamingPolicy

```

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace SecTester.Bus.Internal;
+
+public class JsonHttpMethodEnumerationStringConverter : JsonConverter<HttpMethod>
```
```suggestion
internal class JsonHttpMethodEnumerationStringConverter : JsonConverter<HttpMethod>
```

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026432297
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026437164
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026441039
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026443307
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026454541
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026447321
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026447113
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026447200
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026447472
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026448500
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026446977
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026448358
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026448726
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026448822
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026449070
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026449616
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026450278
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026453400
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026453984

closes #42

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -0,0 +1,74 @@
+  private static readonly HttpMethod _unlock = new("UNLOCK");
+  private static readonly HttpMethod _propfind = new("PROPFIND");
+  private static readonly HttpMethod _view = new("VIEW");
+  private static readonly HttpMethod _trace = new("TRACE");
```
BTW trace is present in `HttpMethod`

derevnjuk commented on test/SecTester.Core.Tests/Utils/StringUtilsTests.cs:
```diff
@@ -0,0 +1,29 @@
+
+public class StringUtilsTests
+{
+  public static readonly IEnumerable<object[]> SnakeCaseMapping = new List<object[]>{
```
```suggestion
  public static readonly IEnumerable<object[]> SnakeCaseMapping = new List<object[]>
  {
```

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026616519
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026616880
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1026422693

closes #42

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -1,55 +1,52 @@
-  private static readonly HttpMethod _trace = new("TRACE");
+  private static readonly IEnumerable<HttpMethod> BaseMethods = typeof(HttpMethod)
+    .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+    .Where(x => x.PropertyType.IsAssignableFrom(typeof(HttpMethod)))
```
`BindingFlags.DeclaredOnly` is responsible for this

```suggestion
```

ostridm replied:
`BindingFlags.DeclaredOnly`  - Specifies that only members declared at the level of the supplied type's hierarchy should be considered. Inherited members are not considered.
Hence it does not perform any type of check and I propose to leave `.Where(x => x.PropertyType.IsAssignableFrom(typeof(HttpMethod)))` just for type casting safety.

derevnjuk replied:
According to the contract, public static members are instances of `HttpMethod`. It is pointless to check it explicitly.

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -1,55 +1,52 @@
+    .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+    .Where(x => x.PropertyType.IsAssignableFrom(typeof(HttpMethod)))
+    .Select(x => x.GetValue(null))
+    .Cast<HttpMethod>()
+    .ToList();
```
```suggestion
    .Cast<HttpMethod>();
```

derevnjuk commented on src/SecTester.Bus/Internal/JsonHttpMethodEnumerationStringConverter.cs:
```diff
@@ -1,55 +1,52 @@
+    new("VIEW")
   };
 
+  private static readonly Dictionary<string, HttpMethod> Map = BaseMethods.Concat(CustomMethods).Distinct()
```
```suggestion
  private static readonly IDictionary<string, HttpMethod> Map = BaseMethods.Concat(CustomMethods).Distinct()
```

derevnjuk replied:
Please use more obvious and clear name

```suggestion
  private static readonly Dictionary<string, HttpMethod> Methods = BaseMethods.Concat(CustomMethods).Distinct()
```

ostridm committed:
feat(scan): define base abstractions to manage scans

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1027068611
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1027068478
https://github.com/NeuraLegion/sectester-net/pull/47#discussion_r1027068419

closes #42
